### ISSUE TITLE
Use `Axes` property

### ISSUE THREAD

**Issue by schriftgestalt** [googlei18n/glyphsLib#280]
There is a custom parameter `Axes`. It contains axis names and tags. It is required for variable font export and does limit the available coordinate fields in the master and instance settings.

**Comment by anthrotype** [comment]
any documentation available yet?

**Comment by anthrotype** [comment]
@schriftgestalt in #279 you also mentioned a "Virtual Master". What's that? You may have already explained this to me in Montreal but I forgot :grimacing:

**Comment by schriftgestalt** [comment]
the format of the Axes parameter:
```
{
	name = Axes;
	value = (
		{
			Name = Weight;
			Tag = wght;
		},
		{
			Name = Width;
			Tag = wdth;
		}
	);
}
```

**Comment by anthrotype** [comment]
and the axis location for each master?

**Comment by schriftgestalt** [comment]
That is defined in the masters settings. They have coordinates for each axis.

**Comment by m4rc1e** [comment]
> It contains axis names and tags. It is required for variable font export

As far as I am aware, this only happens in Glyphs

![screen shot 2017-11-08 at 10 55 45](https://user-images.githubusercontent.com/7525512/32545361-655eed14-c473-11e7-8b38-c99b25c28949.png)


In Fontmake, it will automatically gen axises if non are specified. I love this because I am dealing with many families which do not contain an axis property. Could Fontmake and its supporting libs such as this one still retain this functionality?


Apologies if this is slightly off topic.

**Comment by schriftgestalt** [comment]
Keeping the old mechanism as a fallback might be a good idea. But optimise it to omit "empty" axes.

**Comment by anthrotype** [comment]
@m4rc1e I think we will keep that even as we add support for the new parameters.

@schriftgestalt I mean, the user-space locations for registered axes that have predefined rangee like weight (1-1000) or width (percentage). We are currently getting these from the instances to generate the axes map elements in the designspace (used to build avar table) because in the old Glyphs.app masters only had design-space locations (any arbitary number used by the designers).

**Comment by schriftgestalt** [comment]
There is master parameter for that:
```
{
	name = "Axis Location";
	value = (
		{
			Axis = Weight;
			Location = 100;
		},
		{
			Axis = Width;
			Location = 200;
		}
	);
}
```
You can’t remap Virtual masters. But those are most likely used for non standard axes that don't have a predefined range so it is fine using the internal numbers. (I think of something if someone presents a valid use case).

**Comment by anthrotype** [comment]
Excellent! Thanks again for the clarification.

**Comment by anthrotype** [comment]
@belluzj what about this one? (sorry, I'm trying to close my tabs...)

**Comment by belluzj** [comment]
I'm working on it in #324 :) I had started already in the roundtrip PR but I had understood a few things wrong, so I will try to get it right this time.

@schriftgestalt I have a few questions:
1. If I understand correctly, as soon as there is an "Axes" parameter, the GSFontMaster properties `weightValue`, `widthValue`, `customValue`, `customValue1` etc. will be used in this order to store design locations, regardless of the whether the first defined axis is actually weight, the second is actually width? That's what I got from creating a bunch of test files with the latest Glyphs 2.5
2. What you said before is that user locations are now stored in "Axis Location" for the masters. Is it the same for the instances?
3. If "Axis Location" also works for instances, and when the axis in question is weight or width, should I also read/write the old `weight` and `width` properties? (the ones that match values from a drop-down menu in the instance panel)?

Thanks in advance!

**Comment by schriftgestalt** [comment]
> If I understand correctly, as soon as there is an "Axes" parameter, the GSFontMaster properties weightValue, widthValue, customValue, customValue1 etc. will be used in this order to store design locations, regardless of the whether the first defined axis is actually weight, the second is actually width? That's what I got from creating a bunch of test files with the latest Glyphs 2.5

Correct. Again, that what to keep the noise in the files low and keep it working in older versions.
A future version of Glyphs will clean that up (and the master names, too).

> What you said before is that user locations are now stored in "Axis Location" for the masters. Is it the same for the instances?

>If "Axis Location" also works for instances, and when the axis in question is weight or width, should I also read/write the old weight and width properties? (the ones that match values from a drop-down menu in the instance panel)?

Instances don’t have a `Axis Location`. They have a weight/width class and the `Axis Location` for other axes could be interpolated from the masters?

**Comment by belluzj** [comment]
> Instances don’t have a Axis Location. They have a weight/width class and the Axis Location for other axes could be interpolated from the masters?

OK that makes sense. To make sure I understand:
* the axis "mapping" (the function that goes from user locations to design locations) should be entirely encoded by the masters' (and virtual masters?)' design locations = `xxxValue` and user locations = "Axis Location" parameter
* instances only have a design location and their user location can be found by looking at the axis mapping backwards (using the inverse function of the mapping)
* however, in the specific case of Weight and Width axes, the instances do have a user location, because that's what the weight/width class is, right?

**Comment by schriftgestalt** [comment]
Yes

**Comment by belluzj** [comment]
Thanks a lot for the clarifications. I have two more questions about instance locations along the weight and width axes:

1. what if there's a conflict between the mapping that I get from the masters and the mapping that I get from the instances? Should one of those two mappings take precedence over the other?

Right now, glyphsLib infers the mapping for weight and width from the instances because there are usually more instances than masters (so more points on the mapping). Also in old glyphs files, masters usually don't have proper weightClass values like the instances do, so it's not possible to get a good mapping from masters in that case.

2. Maybe glyphsLib should behave like it does now if the file does not have the "Axes" parameter, and switch to the new system based on the masters if the file has an "Axes" parameter? How does that sound?

I know that we said we should only support the "latest and greatest" from Glyphs (which would be: only support explicit axes with "Axes" and "Axis Location"), but I don't think that everyone will be switching to this system soon, so for a while glyphsLib will have to read both?

**Comment by anthrotype** [comment]
> Maybe glyphsLib should behave like it does now if the file does not have the "Axes" parameter, and switch to the new system based on the masters if the file has an "Axes" parameter?

I think that's reasonable. It would reduce the amount of breaks. We could emit a warning suggesting the user to update their sources.

**Comment by schriftgestalt** [comment]
I have a method `font.axes` that looks for the custom parameter and returned it or, if there is none, it generated one. It checks if the masters are actually any different in each dimension...

**Comment by anthrotype** [comment]
> instances only have a design location and their user location can be found by looking at the axis mapping backwards (using the inverse function of the mapping)
> however, in the specific case of Weight and Width axes, the instances do have a user location, because that's what the weight/width class is, right?

>> Yes

> what if there's a conflict between the mapping that I get from the masters and the mapping that I get from the instances? Should one of those two mappings take precedence over the other?

indeed, what should happen in that case? who wins?

**Comment by anthrotype** [comment]
At least for old files, can we restore the previous code whereby we apply the instances weight/width classes based on the Glyphs.app Weight/Width enumerations (i.e. the drop-down menus in Instances panel, possibly overridden by the weightClass or widthClass instance custom parameters), which we can store in the designspace instance lib so they can be retrieved by the apply_instance_data function? 
I think that'd be better, and easier (the code handling this is getting a bit too complicated). 

The problem is, if we are going to run `apply_instance_data` on any designspaces, even those which may not be created from glyphsLib (like i'm doing in [this fontmake PR 410](https://github.com/googlei18n/fontmake/pull/410/files#diff-77503f3639ebd287f468b3d480600aa9R469), then we should modify the instance UFOs weight/width classes only if the designspace instance lib contain private Glyphs.app instance data, otherwise leave it like that.

For regular UFO+designspace workflows, the weight/width classes are already correctly defined in the UFO masters, and those of the instances are simply interpolated by mutatorMath according to the instance locations. It's only for Glyphs that we had to resort to use the instances' user locations, because before the new Axis Location custom parameter, we could not assign a user space location to the Glyphs masters directly.

**Comment by anthrotype** [comment]
@schriftgestalt We would like an answer to the question Jany asked above in https://github.com/googlei18n/glyphsLib/issues/280#issuecomment-374918169. That is, when there's a conflict between the mapping that we get from the masters and the mapping that we get from the instances, should one of those two mappings take precedence over the other?

Right now in glyphsLib, whenever Axes and Axis Location custom parameters are set, we are only using the masters' locations to define the designspace axes and the mappings from user-space to design-space locations.

Whereas when the new custom parameters are not present, we are using the instances' locations to define these.

The problem is that the instances' user-space locations along the weight/width axis (as defined by the "Weight" and "Width" drop-down menus) could get out of sync with the values that would be interpolated from the instances' location with respect to the masters' user-space locations (as defined in Axis Locations custom parameters).
Not to mention that an instance's weight class can also be arbitrarily overridden with a `weightClass` custom parameter...

**Comment by schriftgestalt** [comment]
In Glyphs, only the master settings are used. I don’t try to compute the wight class from the instances.

**Comment by anthrotype** [comment]
then a named instance in a variable font could have its interpolated weight/width class set differently from a static instance with the same name, generated from the very same source file, if the latter has weight/width class settings inconsistent with those of the masters.

**Comment by schriftgestalt** [comment]
That the wight class of the masters and instances is not in sync is one of the reasons that I still have a "beta" badge on the export settings dialog.
